internal/identity/github.go

```diff
@@ -0,0 +1,170 @@
+package identity
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/pomerium/pomerium/internal/httputil"
+	"github.com/pomerium/pomerium/internal/sessions"
+	"github.com/pomerium/pomerium/internal/version"
+
+	"golang.org/x/oauth2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+const (
+	defaultGitHubProviderURL = "https://github.com"
+	githubAuthURL            = "/login/oauth/authorize"
+	githubUserURL            = "https://api.github.com/user"
+	githubUserGroupURL       = "https://api.github.com/user/orgs"
+	githubRevokeURL          = "https://github.com/oauth/revoke"
+
+	// since github doesn't implement oidc, we need this to refresh the user session
+	refreshDeadline = time.Minute * 60
+)
+
+// GitHubProvider is an implementation of the OAuth Provider.
+type GitHubProvider struct {
+	*Provider
+
+	authURL      string
+	tokenURL     string
+	userEndpoint string
+
+	RevokeURL string `json:"revocation_endpoint"`
+}
+
+// NewGitHubProvider returns a new GitHubProvider.
+func NewGitHubProvider(p *Provider) (*GitHubProvider, error) {
+	gp := &GitHubProvider{
+		authURL:      defaultGitHubProviderURL + githubAuthURL,
+		tokenURL:     defaultGitHubProviderURL + "/login/oauth/access_token",
+		userEndpoint: githubUserURL,
+		RevokeURL:    githubRevokeURL,
+	}
+
+	if p.ProviderURL == "" {
+		p.ProviderURL = defaultGitHubProviderURL
+	}
+
+	if len(p.Scopes) == 0 {
+		p.Scopes = []string{"read:org"}
+	}
+
+	p.oauth = &oauth2.Config{
+		ClientID:     p.ClientID,
+		ClientSecret: p.ClientSecret,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  gp.authURL,
+			TokenURL: gp.tokenURL,
+		},
+		RedirectURL: p.RedirectURL.String(),
+		Scopes:      p.Scopes,
+	}
+	gp.Provider = p
+
+	return gp, nil
+}
+
+// Authenticate creates an identity session with github from a authorization code, and follows up
+// call to the user and user group endpoint with the
+func (p *GitHubProvider) Authenticate(ctx context.Context, code string) (*sessions.State, error) {
+	resp, err := p.oauth.Exchange(ctx, code)
+	if err != nil {
+		return nil, fmt.Errorf("identity/github: token exchange failed %v", err)
+	}
+
+	s := &sessions.State{
+		AccessToken: &oauth2.Token{
+			AccessToken: resp.AccessToken,
+			TokenType:   resp.TokenType,
+		},
+		AccessTokenID: resp.AccessToken,
+	}
+
+	s, err = p.userInfo(ctx, s)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// userInfo will get the user information from github and also retrieve the user's organization(s)
+func (p *GitHubProvider) userInfo(ctx context.Context, s *sessions.State) (*sessions.State, error) {
+	if s == nil || s.AccessToken == nil {
+		return nil, errors.New("identity/github: user session cannot be empty")
+	}
+
+	var response struct {
+		ID        int    `json:"id"`
+		Login     string `json:"login"`
+		Name      string `json:"name"`
+		Email     string `json:"email"`
```

Comment:
It looks like email does not always show up. I'm not sure if it is possible to adjust Oauth2 scopes to do this or not.

<img width="513" alt="Screen Shot 2020-04-06 at 10 51 15 AM" src="https://user-images.githubusercontent.com/1544881/78592194-90e96b80-77f9-11ea-9aa8-fe9f76c4672e.png">

Reply:
the public email will show up but the private email won't, I'll add the `user` scope, this should resolve it.

Reply:
Looks like we have to make _an additional_ call to the user's api.
https://developer.github.com/v3/users/emails/#list-email-addresses-for-a-user

Reply:
sadly, yes!

Reply:
A new update has been made to be able to get private emails, please test it out.